src/ott/solvers/nn/neuraldual.py

```diff
@@ -12,49 +12,78 @@
 """A Jax implementation of the ICNN based Kantorovich dual."""
 
 import warnings
-from typing import Dict, Iterable, List, Optional, Tuple, Union
+from typing import Callable, Dict, Iterable, List, Optional, Tuple, Union
 
 import flax.linen as nn
 import jax
 import jax.numpy as jnp
 import optax
 from flax import core
+from jax.lax import stop_gradient
 from typing_extensions import Literal
 
 from ott.geometry import costs
 from ott.problems.linear import potentials
 from ott.solvers.nn import icnn
+from ott.solvers.nn.conjugate_solver import ConjugateSolverLBFGS
 
 __all__ = ["NeuralDualSolver"]
 
 Train_t = Dict[Literal["train_logs", "valid_logs"], Dict[str, List[float]]]
+Potentials_t = potentials.DualPotentials
+
+default_conjugate_solver = ConjugateSolverLBFGS(
+    gtol=1e-3, max_iter=10, max_linesearch_iter=10
+)
 
 
 class NeuralDualSolver:
```

Comment:
Maybe we should start using a clearer naming convention and drop the `Solver`, since it is not used in for other solvers (e.g. `Sinkhorn` or `GromovWasserstein`). We might also want to specify this is a 2Wasserstein problem. So maybe `2WassersteinDual` ? or  `NeuralDual2Was`? or any other idea : )

Reply:
It could also be clean to let`NeuralDual` be the Euclidean Wasserstein-2 one since it's the most commonly used one, and then give neural OT variations more specialized names.

Slightly related: there's also the question of whether "solver" should remain in the names for the conjugate solver. It may be helpful to keep since it's not always clear the conjugate is computed with an explicit solver, but we could also consider renaming to something like `ConjugateLBFGS`

Reply:
Maybe `W2NeuralDual` then? (we can expect someone will implement `W1` at some point)

I also think `Fenchel` or `Legendre` is a bit more explicit than `Conjugate` in convex analysis literature. Maybe introduce `Fenchel` and then use the solver by default being LBFGS? 

Another question I am wondering is, wouldnt it be a bit more "JAX functional" to have a function transform taking a function `f` in and returning another function `tilde_f`?

Reply:
I updated `ConjugateSolverLBFGS->FenchelConjugateLBFGS` in https://github.com/ott-jax/ott/pull/219/commits/0681a034ee797d6be4dc13213133f6bb65e6b7d1 but kept the argument name to `W2NeuralDual` as `conjugate_solver` for now. Do you think we should update this name too?

> Another question I am wondering is, wouldnt it be a bit more "JAX functional" to have a function transform taking a function f in and returning another function tilde_f?

One way this could be done is by making the solver a module that takes the function as the input and then updating `solve` to `__call__`. I originally didn't do it this way since I didn't want to construct a new function/object every time to compute the conjugate, but maybe it's not so bad and isn't much overhead